# Q1. Write a Python function that takes in two arrays of data and calculates the F-value for a variance ratio test. The function should return the F-value and the corresponding p-value for the test.

In [1]:
import numpy as np
from scipy.stats import f

def variance_ratio_test(x, y):
    """
    Perform a variance ratio test to compare the variances of two data arrays x and y.
    
    Parameters:
    x (array-like): First data array.
    y (array-like): Second data array.
    
    Returns:
    F (float): F-value for the variance ratio test.
    p_value (float): Corresponding p-value for the test.
    """
    n_x = len(x)
    n_y = len(y)
    var_x = np.var(x, ddof=1)
    var_y = np.var(y, ddof=1)
    
    if var_x > var_y:
        F = var_x / var_y
        df1 = n_x - 1
        df2 = n_y - 1
    else:
        F = var_y / var_x
        df1 = n_y - 1
        df2 = n_x - 1
    
    p_value = f.sf(F, df1, df2)
    
    return F, p_value

In [2]:
x = [1, 2, 3, 4, 5]
y = [2, 4, 6, 8, 10]

F, p_value = variance_ratio_test(x, y)

print("F-value:", F)
print("p-value:", p_value)

F-value: 4.0
p-value: 0.10400000000000002


# Q2. Given a significance level of 0.05 and the degrees of freedom for the numerator and denominator of an F-distribution, write a Python function that returns the critical F-value for a two-tailed test.

In [3]:
from scipy.stats import f

def critical_f_value(dfn,dfd,alpha):
    """
    Calculate the critical F-value for a two-tailed test given degrees of freedom for numerator and denominator of an F-distribution and a significance level.
    
    Parameters:
    dfn (int): Degrees of freedom for the numerator.
    dfd (int): Degrees of freedom for the denominator.
    alpha (float): Significance level.
    
    Returns:
    crit_f (float): Critical F-value for a two-tailed test.
    """
    crit_f = f.ppf(1 - alpha / 2, dfn, dfd)
    return crit_f

In [4]:
dfn = 3
dfd = 10
alpha = 0.05

crit_f = critical_f_value(dfn, dfd, alpha)

print("Critical F-value:", crit_f)

Critical F-value: 4.825621493405406


# Q3. Write a Python program that generates random samples from two normal distributions with known variances and uses an F-test to determine if the variances are equal. The program should output the F-value, degrees of freedom, and p-value for the test.

In [8]:
import numpy as np
from scipy.stats import f_oneway

# Set the seed for reproducibility
np.random.seed(123)

# Generate two normal distributions with known variances
mean1 = 10
mean2 = 12
var1 = 4
var2 = 9

n1 = 20
n2 = 30

sample1 = np.random.normal(mean1, np.sqrt(var1), n1)
sample2 = np.random.normal(mean2, np.sqrt(var2), n2)

# Calculate the F-value, degrees of freedom, and p-value for the F-test
F, p_value = f_oneway(sample1, sample2)

df1 = n1 - 1
df2 = n2 - 1

# Print the results
print("Sample 1 mean:", np.mean(sample1))
print("Sample 1 variance:", np.var(sample1, ddof=1))
print("Sample 2 mean:", np.mean(sample2))
print("Sample 2 variance:", np.var(sample2, ddof=1))

print("F-value:", F)
print("Degrees of freedom:", df1, ",", df2)
print("p-value:", p_value)

Sample 1 mean: 10.228835463910581
Sample 1 variance: 6.324748987421546
Sample 2 mean: 11.837521741508342
Sample 2 variance: 12.530651965031726
F-value: 3.0825888511393837
Degrees of freedom: 19 , 29
p-value: 0.0855142060768227


# Q4.The variances of two populations are known to be 10 and 15. A sample of 12 observations is taken from each population. Conduct an F-test at the 5% significance level to determine if the variances are significantly different.

# Ans. :

To conduct an F-test to determine if the variances of two populations are significantly different, we need to set up the null and alternative hypotheses:

* __Null hypothesis:__ The variances of the two populations are equal.
* __Alternative hypothesis:__ The variances of the two populations are not equal.
We will use a significance level of 0.05 for this test.

The F-test statistic is calculated as the ratio of the larger sample variance to the smaller sample variance. In this case, the larger sample variance is 15 and the smaller sample variance is 10, so the F-test statistic is:

__F = 15/10 = 1.5__

To calculate the p-value for this test, we need to look up the critical F-value in an F-distribution table or use a calculator. With 11 degrees of freedom for the numerator and 11 degrees of freedom for the denominator (since we have 12 observations in each sample), the critical F-value at a 5% significance level is approximately 2.687.

Since the calculated F-value of 1.5 is less than the critical F-value of 2.687, we fail to reject the null hypothesis. We do not have sufficient evidence to conclude that the variances of the two populations are significantly different at the 5% significance level.

Therefore, we can conclude that there is not enough evidence to suggest that the variances of the two populations are different at a 5% significance level.

# Q5. A manufacturer claims that the variance of the diameter of a certain product is 0.005. A sample of 25 products is taken, and the sample variance is found to be 0.006. Conduct an F-test at the 1% significance level to determine if the claim is justified.

# Ans. 

To conduct an F-test to determine if the manufacturer's claim about the variance of the diameter of the product is justified, we need to set up the null and alternative hypotheses:

* __Null hypothesis:__ The variance of the diameter of the product is equal to 0.005.
* __Alternative hypothesis:__ The variance of the diameter of the product is not equal to 0.005.
We will use a significance level of 0.01 for this test.

The F-test statistic is calculated as the ratio of the sample variance to the hypothesized population variance. In this case, the sample variance is 0.006 and the hypothesized population variance is 0.005, so the F-test statistic is:

__F = 0.006/0.005 = 1.2__

To calculate the p-value for this test, we need to look up the critical F-value in an F-distribution table or use a calculator. With 24 degrees of freedom for the numerator (since we have 25 observations in the sample) and 1 degree of freedom for the denominator, the critical F-value at a 1% significance level is approximately 6.60.

Since the calculated F-value of 1.2 is less than the critical F-value of 6.60, we fail to reject the null hypothesis. We do not have sufficient evidence to conclude that the variance of the diameter of the product is significantly different from 0.005 at the 1% significance level.

Therefore, we can conclude that the manufacturer's claim about the variance of the diameter of the product is justified at a 1% significance level.

# Q6. Write a Python function that takes in the degrees of freedom for the numerator and denominator of an F-distribution and calculates the mean and variance of the distribution. The function should return the mean and variance as a tuple.

In [10]:
import numpy as np

def f_distribution_stats(df_num, df_den):
    mean = df_den / (df_den - 2)
    var = (2 * (df_den**2) * (df_num + df_den - 2)) / ((df_num * (df_den - 2)**2 * (df_den - 4)))
    return mean, var

In [11]:
mean, var = f_distribution_stats(5, 10)
print("Mean:", mean)
print("Variance:", var)

Mean: 1.25
Variance: 1.3541666666666667


# Q7. A random sample of 10 measurements is taken from a normal population with unknown variance. The sample variance is found to be 25. Another random sample of 15 measurements is taken from another normal population with unknown variance, and the sample variance is found to be 20. Conduct an F-test at the 10% significance level to determine if the variances are significantly different.

# Ans. :

To conduct an F-test to determine if the variances of the two populations are significantly different, we need to set up the null and alternative hypotheses:

* __Null hypothesis:__ The variances of the two populations are equal.
* __Alternative hypothesis:__ The variances of the two populations are not equal.
We will use a significance level of 0.10 for this test.

The F-test statistic is calculated as the ratio of the larger sample variance to the smaller sample variance. In this case, the larger sample variance is 25 and the smaller sample variance is 20, so the F-test statistic is:

__F = 25/20 = 1.25__

To calculate the p-value for this test, we need to look up the critical F-value in an F-distribution table or use a calculator. With 9 degrees of freedom for the numerator (since we have 10 observations in the first sample) and 14 degrees of freedom for the denominator (since we have 15 observations in the second sample), the critical F-value at a 10% significance level is approximately 2.50.

Since the calculated F-value of 1.25 is less than the critical F-value of 2.50, we fail to reject the null hypothesis. We do not have sufficient evidence to conclude that the variances of the two populations are significantly different at the 10% significance level.

Therefore, we can conclude that there is no significant difference between the variances of the two populations at a 10% significance level.

# Q8. The following data represent the waiting times in minutes at two different restaurants on a Saturday night: Restaurant A: 24, 25, 28, 23, 22, 20, 27; Restaurant B: 31, 33, 35, 30, 32, 36. Conduct an F-test at the 5% significance level to determine if the variances are significantly different.

# Ans. :

To conduct an F-test to determine if the variances of the waiting times at two different restaurants are significantly different, we need to set up the null and alternative hypotheses:

* __Null hypothesis:__ The variances of the waiting times at the two restaurants are equal.
* __Alternative hypothesis:__ The variances of the waiting times at the two restaurants are not equal.<br>

We will use a significance level of 0.05 for this test.

First, we need to calculate the sample variances for each restaurant:

* Sample variance for Restaurant A: sA^2 = ((24-24.14)^2 + (25-24.14)^2 + (28-24.14)^2 + (23-24.14)^2 + (22-24.14)^2 + (20-24.14)^2 + (27-24.14)^2)/6 = 8.61
* Sample variance for Restaurant B: sB^2 = ((31-32.83)^2 + (33-32.83)^2 + (35-32.83)^2 + (30-32.83)^2 + (32-32.83)^2 + (36-32.83)^2)/5 = 7.97<br>

Next, we calculate the F-test statistic, which is the ratio of the larger sample variance to the smaller sample variance:

__F = sA^2 / sB^2 = 8.61 / 7.97 = 1.08__

To calculate the p-value for this test, we need to look up the critical F-value in an F-distribution table or use a calculator. With 6 degrees of freedom for the numerator (since we have 7 observations in Restaurant A) and 5 degrees of freedom for the denominator (since we have 6 observations in Restaurant B), the critical F-value at a 5% significance level is approximately 4.39.

Since the calculated F-value of 1.08 is less than the critical F-value of 4.39, we fail to reject the null hypothesis. We do not have sufficient evidence to conclude that the variances of the waiting times at the two restaurants are significantly different at the 5% significance level.

Therefore, we can conclude that there is no significant difference between the variances of the waiting times at the two restaurants at a 5% significance level.

# Q9. The following data represent the test scores of two groups of students: Group A: 80, 85, 90, 92, 87, 83; Group B: 75, 78, 82, 79, 81, 84. Conduct an F-test at the 1% significance level to determine if the variances are significantly different.

# Ans. :


To conduct an F-test to determine if the variances of the two groups are significantly different, we need to set up the null and alternative hypotheses:

* __Null hypothesis:__ The variances of the two groups are equal.
* __Alternative hypothesis:__ The variances of the two groups are not equal.
We will use a significance level of 0.01 for this test.

First, we need to calculate the sample variances for each group:

* Sample variance for Group A: sA^2 = ((80-87.17)^2 + (85-87.17)^2 + (90-87.17)^2 + (92-87.17)^2 + (87-87.17)^2 + (83-87.17)^2)/5 = 19.53
* Sample variance for Group B: sB^2 = ((75-80.83)^2 + (78-80.83)^2 + (82-80.83)^2 + (79-80.83)^2 + (81-80.83)^2 + (84-80.83)^2)/5 = 8.31<br>

Next, we calculate the F-test statistic, which is the ratio of the larger sample variance to the smaller sample variance:

__F = sA^2 / sB^2 = 19.53 / 8.31 = 2.35__

To calculate the p-value for this test, we need to look up the critical F-value in an F-distribution table or use a calculator. With 5 degrees of freedom for the numerator (since we have 6 observations in Group A) and 5 degrees of freedom for the denominator (since we have 6 observations in Group B), the critical F-value at a 1% significance level is approximately 7.71.

Since the calculated F-value of 2.35 is less than the critical F-value of 7.71, we fail to reject the null hypothesis. We do not have sufficient evidence to conclude that the variances of the two groups are significantly different at the 1% significance level.

Therefore, we can conclude that there is no significant difference between the variances of the two groups at a 1% significance level.